### ISSUE TITLE
Use-after-free of a COrder_Resource on victory

### ISSUE THREAD

**Issue by Ace17** [Wargus/stratagus#400]
Happens immediately when clicking on the "Victory" button.
Below, it's detected using asan:

```
==17735==ERROR: AddressSanitizer: heap-use-after-free on address 0x60800029b7b8 at pc 0x555b40556b63 bp 0x7ffe39949840 sp 0x7ffe39949838
WRITE of size 8 at 0x60800029b7b8 thread T0
    #0 0x555b40556b62 in CUnitPtr::Reset() /tmp/dev/stratagus/src/unit/unitptr.cpp:56
    #1 0x555b3fa5446b in CUnitPtr::~CUnitPtr() /tmp/dev/stratagus/src/include/unitptr.h:47
    #2 0x555b3fabfe28 in COrder_Resource::~COrder_Resource() /tmp/dev/stratagus/src/action/action_resource.cpp:241
    #3 0x555b3fabff9b in COrder_Resource::~COrder_Resource() /tmp/dev/stratagus/src/action/action_resource.cpp:241
    #4 0x555b404bdba2 in CUnit::Release(bool) /tmp/dev/stratagus/src/unit/unit.cpp:561
    #5 0x555b404ff571 in CleanUnits() /tmp/dev/stratagus/src/unit/unit.cpp:3500
    #6 0x555b3fcd07b4 in CleanGame() /tmp/dev/stratagus/src/game/game.cpp:1119
    #7 0x555b3fcbb49b in StartMap(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool) /tmp/dev/stratagus/src/game/game.cpp:156
    #8 0x555b4071df61 in tolua_stratagus_StartMap00 /tmp/dev/stratagus/bin/tolua.cpp:1099
    #9 0x7feda4082118  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf118)
    #10 0x7feda408dab6  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0x1aab6)
    #11 0x7feda40827a4  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf7a4)
    #12 0x7feda4081a9a  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xea9a)
    #13 0x7feda408294f  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf94f)
    #14 0x7feda407dd77 in lua_pcall (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xad77)
    #15 0x555b40210a5d in LuaCall(lua_State*, int, int, int, bool) /tmp/dev/stratagus/src/stratagus/script.cpp:203
    #16 0x555b401e1976 in LuaCallback::run(int) /tmp/dev/stratagus/src/stratagus/luacallback.cpp:180
    #17 0x555b403828bc in LuaActionListener::action(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /tmp/dev/stratagus/src/ui/widgets.cpp:196
    #18 0x555b3fd78bf0 in gcn::Widget::generateAction() /tmp/dev/stratagus/src/guichan/widget.cpp:599
    #19 0x555b3fd99471 in gcn::Button::mouseClick(int, int, int, int) /tmp/dev/stratagus/src/guichan/widgets/button.cpp:236
    #20 0x555b3fd7374b in gcn::Widget::_mouseInputMessage(gcn::MouseInput const&) /tmp/dev/stratagus/src/guichan/widget.cpp:469
    #21 0x555b3fda9df3 in gcn::Container::_mouseInputMessage(gcn::MouseInput const&) /tmp/dev/stratagus/src/guichan/widgets/container.cpp:373
    #22 0x555b3fd269aa in gcn::Gui::logic() /tmp/dev/stratagus/src/guichan/gui.cpp:155
    #23 0x555b4038179e in handleInput(SDL_Event const*) /tmp/dev/stratagus/src/ui/widgets.cpp:155
    #24 0x555b406364cf in WaitEventsOneFrame() /tmp/dev/stratagus/src/video/sdl.cpp:809
    #25 0x555b4040211b in MenuScreen::run(bool) /tmp/dev/stratagus/src/ui/widgets.cpp:2427
    #26 0x555b4078fa82 in tolua_stratagus_CMenuScreen_run00 /tmp/dev/stratagus/bin/tolua.cpp:21254
    #27 0x7feda4082118  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf118)
    #28 0x7feda408dab6  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0x1aab6)
    #29 0x7feda40827a4  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf7a4)
    #30 0x7feda4081a9a  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xea9a)
    #31 0x7feda408294f  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf94f)
    #32 0x7feda407dd77 in lua_pcall (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xad77)
    #33 0x555b40210a5d in LuaCall(lua_State*, int, int, int, bool) /tmp/dev/stratagus/src/stratagus/script.cpp:203
    #34 0x555b402109e7 in LuaCall(int, int, bool) /tmp/dev/stratagus/src/stratagus/script.cpp:175
    #35 0x555b402114b7 in LuaLoadFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool) /tmp/dev/stratagus/src/stratagus/script.cpp:274
    #36 0x555b40211a84 in CclLoad /tmp/dev/stratagus/src/stratagus/script.cpp:311
    #37 0x7feda4082118  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf118)
    #38 0x7feda408dab6  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0x1aab6)
    #39 0x7feda40827a4  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf7a4)
    #40 0x7feda4081a9a  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xea9a)
    #41 0x7feda408294f  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf94f)
    #42 0x7feda407dd77 in lua_pcall (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xad77)
    #43 0x555b40210a5d in LuaCall(lua_State*, int, int, int, bool) /tmp/dev/stratagus/src/stratagus/script.cpp:203
    #44 0x555b401e1976 in LuaCallback::run(int) /tmp/dev/stratagus/src/stratagus/luacallback.cpp:180
    #45 0x555b403853fd in LuaActionListener::mouseClick(int, int, int, int) /tmp/dev/stratagus/src/ui/widgets.cpp:260
    #46 0x555b3fd7374b in gcn::Widget::_mouseInputMessage(gcn::MouseInput const&) /tmp/dev/stratagus/src/guichan/widget.cpp:469
    #47 0x555b3fe29aa4 in gcn::ScrollArea::_mouseInputMessage(gcn::MouseInput const&) /tmp/dev/stratagus/src/guichan/widgets/scrollarea.cpp:313
    #48 0x555b3fda9df3 in gcn::Container::_mouseInputMessage(gcn::MouseInput const&) /tmp/dev/stratagus/src/guichan/widgets/container.cpp:373
    #49 0x555b3fd269aa in gcn::Gui::logic() /tmp/dev/stratagus/src/guichan/gui.cpp:155
    #50 0x555b4038179e in handleInput(SDL_Event const*) /tmp/dev/stratagus/src/ui/widgets.cpp:155
    #51 0x555b406364cf in WaitEventsOneFrame() /tmp/dev/stratagus/src/video/sdl.cpp:809
    #52 0x555b4040211b in MenuScreen::run(bool) /tmp/dev/stratagus/src/ui/widgets.cpp:2427
    #53 0x555b4078fa82 in tolua_stratagus_CMenuScreen_run00 /tmp/dev/stratagus/bin/tolua.cpp:21254
    #54 0x7feda4082118  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf118)
    #55 0x7feda408dab6  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0x1aab6)
    #56 0x7feda40827a4  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf7a4)
    #57 0x7feda4081a9a  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xea9a)
    #58 0x7feda408294f  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf94f)
    #59 0x7feda407dd77 in lua_pcall (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xad77)
    #60 0x555b40210a5d in LuaCall(lua_State*, int, int, int, bool) /tmp/dev/stratagus/src/stratagus/script.cpp:203
    #61 0x555b401e1976 in LuaCallback::run(int) /tmp/dev/stratagus/src/stratagus/luacallback.cpp:180
    #62 0x555b403828bc in LuaActionListener::action(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /tmp/dev/stratagus/src/ui/widgets.cpp:196
    #63 0x555b3fd78bf0 in gcn::Widget::generateAction() /tmp/dev/stratagus/src/guichan/widget.cpp:599
    #64 0x555b3fd99471 in gcn::Button::mouseClick(int, int, int, int) /tmp/dev/stratagus/src/guichan/widgets/button.cpp:236
    #65 0x555b3fd7374b in gcn::Widget::_mouseInputMessage(gcn::MouseInput const&) /tmp/dev/stratagus/src/guichan/widget.cpp:469
    #66 0x555b3fda9df3 in gcn::Container::_mouseInputMessage(gcn::MouseInput const&) /tmp/dev/stratagus/src/guichan/widgets/container.cpp:373
    #67 0x555b3fd269aa in gcn::Gui::logic() /tmp/dev/stratagus/src/guichan/gui.cpp:155
    #68 0x555b4038179e in handleInput(SDL_Event const*) /tmp/dev/stratagus/src/ui/widgets.cpp:155
    #69 0x555b406364cf in WaitEventsOneFrame() /tmp/dev/stratagus/src/video/sdl.cpp:809
    #70 0x555b4040211b in MenuScreen::run(bool) /tmp/dev/stratagus/src/ui/widgets.cpp:2427
    #71 0x555b4078fa82 in tolua_stratagus_CMenuScreen_run00 /tmp/dev/stratagus/bin/tolua.cpp:21254
    #72 0x7feda4082118  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf118)
    #73 0x7feda408dc86  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0x1ac86)
    #74 0x7feda40827a4  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf7a4)
    #75 0x7feda4081a9a  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xea9a)
    #76 0x7feda408294f  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf94f)
    #77 0x7feda407dd77 in lua_pcall (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xad77)
    #78 0x555b40210a5d in LuaCall(lua_State*, int, int, int, bool) /tmp/dev/stratagus/src/stratagus/script.cpp:203
    #79 0x555b401e1976 in LuaCallback::run(int) /tmp/dev/stratagus/src/stratagus/luacallback.cpp:180
    #80 0x555b403828bc in LuaActionListener::action(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /tmp/dev/stratagus/src/ui/widgets.cpp:196
    #81 0x555b3fd78bf0 in gcn::Widget::generateAction() /tmp/dev/stratagus/src/guichan/widget.cpp:599
    #82 0x555b3fd99471 in gcn::Button::mouseClick(int, int, int, int) /tmp/dev/stratagus/src/guichan/widgets/button.cpp:236
    #83 0x555b3fd7374b in gcn::Widget::_mouseInputMessage(gcn::MouseInput const&) /tmp/dev/stratagus/src/guichan/widget.cpp:469
    #84 0x555b3fda9df3 in gcn::Container::_mouseInputMessage(gcn::MouseInput const&) /tmp/dev/stratagus/src/guichan/widgets/container.cpp:373
    #85 0x555b3fd269aa in gcn::Gui::logic() /tmp/dev/stratagus/src/guichan/gui.cpp:155
    #86 0x555b4038179e in handleInput(SDL_Event const*) /tmp/dev/stratagus/src/ui/widgets.cpp:155
    #87 0x555b406364cf in WaitEventsOneFrame() /tmp/dev/stratagus/src/video/sdl.cpp:809
    #88 0x555b4040211b in MenuScreen::run(bool) /tmp/dev/stratagus/src/ui/widgets.cpp:2427
    #89 0x555b4078fa82 in tolua_stratagus_CMenuScreen_run00 /tmp/dev/stratagus/bin/tolua.cpp:21254
    #90 0x7feda4082118  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf118)
    #91 0x7feda408dc86  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0x1ac86)
    #92 0x7feda40827a4  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf7a4)
    #93 0x7feda4081a9a  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xea9a)
    #94 0x7feda408294f  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf94f)
    #95 0x7feda407dd77 in lua_pcall (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xad77)
    #96 0x555b40210a5d in LuaCall(lua_State*, int, int, int, bool) /tmp/dev/stratagus/src/stratagus/script.cpp:203
    #97 0x555b402109e7 in LuaCall(int, int, bool) /tmp/dev/stratagus/src/stratagus/script.cpp:175
    #98 0x555b402114b7 in LuaLoadFile(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool) /tmp/dev/stratagus/src/stratagus/script.cpp:274
    #99 0x555b40265a3d in MenuLoop /tmp/dev/stratagus/src/stratagus/stratagus.cpp:313
    #100 0x555b40268b61 in stratagusMain(int, char**) /tmp/dev/stratagus/src/stratagus/stratagus.cpp:788
    #101 0x555b401e2223 in main /tmp/dev/stratagus/src/stratagus/main.cpp:37
    #102 0x7feda264a7fc in __libc_start_main ../csu/libc-start.c:332
    #103 0x555b3fa38409 in _start (/tmp/dev/stratagus/bin/stratagus+0x190a409)

0x60800029b7b8 is located 24 bytes inside of 88-byte region [0x60800029b7a0,0x60800029b7f8)
freed by thread T0 here:
    #0 0x7feda36b0db7 in operator delete(void*, unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cpp:172
    #1 0x555b3fabffac in COrder_Resource::~COrder_Resource() /tmp/dev/stratagus/src/action/action_resource.cpp:241
    #2 0x555b404bdba2 in CUnit::Release(bool) /tmp/dev/stratagus/src/unit/unit.cpp:561
    #3 0x555b404b99cc in CUnit::RefsDecrease() /tmp/dev/stratagus/src/unit/unit.cpp:395
    #4 0x555b40556b19 in CUnitPtr::Reset() /tmp/dev/stratagus/src/unit/unitptr.cpp:54
    #5 0x555b3fa5446b in CUnitPtr::~CUnitPtr() /tmp/dev/stratagus/src/include/unitptr.h:47
    #6 0x555b3fabfe28 in COrder_Resource::~COrder_Resource() /tmp/dev/stratagus/src/action/action_resource.cpp:241
    #7 0x555b3fabff9b in COrder_Resource::~COrder_Resource() /tmp/dev/stratagus/src/action/action_resource.cpp:241
    #8 0x555b404bdba2 in CUnit::Release(bool) /tmp/dev/stratagus/src/unit/unit.cpp:561
    #9 0x555b404ff571 in CleanUnits() /tmp/dev/stratagus/src/unit/unit.cpp:3500
    #10 0x555b3fcd07b4 in CleanGame() /tmp/dev/stratagus/src/game/game.cpp:1119
    #11 0x555b3fcbb49b in StartMap(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool) /tmp/dev/stratagus/src/game/game.cpp:156
    #12 0x555b4071df61 in tolua_stratagus_StartMap00 /tmp/dev/stratagus/bin/tolua.cpp:1099
    #13 0x7feda4082118  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf118)

previously allocated by thread T0 here:
    #0 0x7feda36aff37 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x555b3fae605e in COrder_Resource::Clone() const /tmp/dev/stratagus/src/include/action/action_resource.h:53
    #2 0x555b3fba4af7 in AiMoveUnitInTheWay /tmp/dev/stratagus/src/ai/ai.cpp:906
    #3 0x555b3fba53c3 in AiCanNotMove(CUnit&) /tmp/dev/stratagus/src/ai/ai.cpp:931
    #4 0x555b3fac6fc9 in COrder_Resource::MoveToResource_Unit(CUnit&) /tmp/dev/stratagus/src/action/action_resource.cpp:450
    #5 0x555b3fac78ed in COrder_Resource::MoveToResource(CUnit&) /tmp/dev/stratagus/src/action/action_resource.cpp:477
    #6 0x555b3fae0827 in COrder_Resource::Execute(CUnit&) /tmp/dev/stratagus/src/action/action_resource.cpp:1301
    #7 0x555b3fb280d8 in HandleUnitAction /tmp/dev/stratagus/src/action/actions.cpp:427
    #8 0x555b3fb2b05e in UnitActionsEachCycle<__gnu_cxx::__normal_iterator<CUnit**, std::vector<CUnit*> > > /tmp/dev/stratagus/src/action/actions.cpp:561
    #9 0x555b3fb2963e in UnitActions() /tmp/dev/stratagus/src/action/actions.cpp:599
    #10 0x555b401e4868 in GameLogicLoop /tmp/dev/stratagus/src/stratagus/mainloop.cpp:264
    #11 0x555b401e5349 in SingleGameLoop /tmp/dev/stratagus/src/stratagus/mainloop.cpp:389
    #12 0x555b401e5613 in GameMainLoop() /tmp/dev/stratagus/src/stratagus/mainloop.cpp:429
    #13 0x555b3fcbb482 in StartMap(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool) /tmp/dev/stratagus/src/game/game.cpp:150
    #14 0x555b4071df61 in tolua_stratagus_StartMap00 /tmp/dev/stratagus/bin/tolua.cpp:1099
    #15 0x7feda4082118  (/usr/lib/x86_64-linux-gnu/liblua5.1.so.0+0xf118)

SUMMARY: AddressSanitizer: heap-use-after-free /tmp/dev/stratagus/src/unit/unitptr.cpp:56 in CUnitPtr::Reset()
Shadow bytes around the buggy address:
  0x0c108004b6a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c108004b6b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c108004b6c0: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c108004b6d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c108004b6e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c108004b6f0: fa fa fa fa fd fd fd[fd]fd fd fd fd fd fd fd fa
  0x0c108004b700: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c108004b710: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c108004b720: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c108004b730: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c108004b740: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==17735==ABORTING

```

**Comment by timfel** [comment]
Thank you for the report. Fixed with [94a6d9e](https://github.com/Wargus/stratagus/commit/94a6d9ecd9b42f6f08bef3701e96a99c32598b42). Keep them coming :)

**Comment by Ace17** [comment]
>  Keep them coming :)

Zog zog :)
